lib/autoproj/daemon/github_watcher.rb

```diff
@@ -1,189 +1,224 @@
+                packages.find do |pkg|
+                    pkg.owner == push_event.owner &&
+                        pkg.name == push_event.name &&
+                        pkg.branch == push_event.branch
+                end
+            end
+
+            # @param [Github::PushEvent] push_event
+            # @return [void]
+            def dispatch_push_event(push_event)
+                to_mainline = to_mainline?(
+                    push_event.owner, push_event.name, push_event.branch
+                )
+                if to_mainline
+                    package = package_by_push_event(push_event)
+                    return if package.head_sha == push_event.head_sha
+                else
+                    cached_pull_request = cache.pull_requests.find do |pr|
+                        pr.head_owner == push_event.owner &&
+                            pr.head_name == push_event.name &&
+                            pr.head_branch == push_event.branch
                     end
-                end.size
+                    return unless cached_pull_request
+
+                    pull_request = client.pull_requests(
+                        cached_pull_request.base_owner,
+                        cached_pull_request.base_name,
+                    ).find { |pr| pr.number == cached_pull_request.number }
+                    return unless pull_request
+                end
+                call_push_hooks(push_event, mainline: to_mainline,
+                                            pull_request: pull_request)
             end
-            # rubocop: enable Metrics/AbcSize
 
-            # Whether an event is valid for a given repository
-            # A valid event is either a Pull Request or Push that
-            # are newer than the last handled event
-            #
-            # @return [Boolean]
-            def valid_event?(repo, event)
-                event_types = %w[PullRequestEvent PushEvent]
-                return false unless event_types.any? { |e| e == event['type'] }
-                return false if Time.parse(event['created_at'].to_s) < repo.timestamp
+            # @param [Array] events An array of events
+            # @return [void]
+            def handle_owner_events(events)
+                events = filter_events(events)
+                events = partition_events_by_repo_name(events)
+                events.each_value do |repo_events|
+                    push_events, pull_request_events = partition_events_by_type(
+                        repo_events
+                    )
 
-                true
+                    last_push_event = push_events.max_by(&:created_at)
+                    last_pull_request_event =
+                        pull_request_events.max_by(&:created_at)
+
+                    dispatch_push_event(last_push_event) if last_push_event
+                    next unless last_pull_request_event
+
+                    pr = last_pull_request_event.pull_request
```

Comment:
it would look clearer (for symmetry) that this filtering logic ends up in a separate `dispatch_pull_request_event` method